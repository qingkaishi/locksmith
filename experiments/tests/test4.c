#include <pthread.h>
#include <stdio.h>

int f(pthread_mutex_t *l) {
  int k, *j;
  
  pthread_mutex_init(l, NULL);
  pthread_mutex_lock(l);

  j = &k;
  if(pthread_mutex_trylock(l)) {
    *j = 5;
  }
  else {
    j = (int *) &l;
  }
  pthread_mutex_unlock(l);
  return 2;
}

void *dostuff(void*);

int main() {
  int i = 1;
	pthread_t t1;

	for(i = 0; i<10; i++) { // init i=0 not shared
		pthread_create(&t1, NULL, dostuff, &i);
	}
}

void *dostuff(void*);

void * dostuff(void * a) {
	int *i = (int*) a;
  pthread_mutex_t t;
  printf("foo");
  *i = 0;
  f(&t);
}
