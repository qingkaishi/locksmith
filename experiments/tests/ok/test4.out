************************* STARTING *************************


test4.c:15: Warning: test4.c:15: Probably using char* where void* is best used:
 lock(*l:test4.c:4)*^{l:test4.c:4}*^{&l:test4.c:4} flows to int*^{j:test4.c:5}.  Losing precision.


test4.c:15: Warning: assigning incompatible pointers, conflating flow

test4.c:15: Warning: lock is conflated! (*l:test4.c:4)

test4.c:15: Warning: assigning number to pointer
functions declared and used but not defined:
  printf
shared: &i:test4.c:24

Warning: Possible data race: &i:test4.c:24 is not protected!
 references:
  dereference of &i:test4.c:24 at test4.c:27
    &i:test4.c:24
  locks acquired:
    <empty>
  in: main at test4.c:23 

  dereference of i:test4.c:35 at test4.c:38
    &i:test4.c:24 => cast:test4.c:35 at dostuff test4.c:28
                  => i:test4.c:35
  locks acquired:
    <empty>
  in: main at test4.c:23 


*************************** DONE ***************************
